### DISCUSSION TITLE
libvips C++ caching/reuse

### DISCUSSION THREAD

**Discussion by breunigs** [libvips/libvips#2784] [General]
I am trying to understand libvips' caching behaviour of some commands that leads to unexpected results for me if I cache or reuse some intermediate steps. Since I am new to C++, it's likely that I might have a misconception of how things work.

Here's an abbreviated version of what I do, without any caching/reuse. The non-code description is "blur faces in a video".
This works as expected:

```cpp
for (auto frame : frames)
{
  auto mask = vips::VImage::black(1920, 1080);
  for (auto &[x, y, w, h] : blurs)
  {
    auto tinyMask = vips::VImage::black(w, h)
                        .invert()
                        .embed(2 * blurRadius, 2 * blurRadius, w + 4 * blurRadius, h + 4 * blurRadius,
                               vips::VImage::option()->set("background", std::vector<double>{0.0}))
                        .gaussblur(blurRadius);
    auto mode = vips::VImage::option()->set("mode", VipsCombineMode::VIPS_COMBINE_MODE_ADD);
    mask.draw_image(tinyMask, x, y, mode);
  }

  auto blurred = frame.gaussblur(10.0);
  auto result = mask.ifthenelse(blurred, frame, vips::VImage::option()->set("blend", true));
  result.write(…);
}
```

Since `tinyMask` only depends on the width and height, it has a good cache hit ratio. Therefore I wanted to reuse it instead of calculating it every time.

I won't include the code for that, but in principle I stick it into a `std::map<std::tuple<int, int>, vips::VImage>`. This results in the `mask` for frame 2 to also include the ones from frame 1, or at least the ones that had a successful cache hit.

My initial thought was that `draw_image` is modifying the cached `tinyMask` in some fashion, but calling `tinyMask.copy()` or `tinyMask.copy_memory()` made no difference. Through chance I tried `mask = vips::VImage::black(1920, 1080).copy();` (i.e. added a `copy()`), which solves the issue. For me that would imply that the result of `vips::VImage::black(…)` is cached and that `draw_image` makes destructive changes to it. However, with that assumption it should also have triggered before I cached the `tinyMask`s, so it can't be true. 

Any hints on where I made some wrong assumptions, or maybe some docs I can read on this topic?

Thanks,
Stefan

**Comment by jcupitt** [comment]
Hello @breunigs,

The draw operations are odd. This line will create a partial (virtual?) image:

```C++
  auto mask = vips::VImage::black(1920, 1080);
```

The first time you draw on it, the draw operation will have to turn it into a memory image, so this line:

```C++
    mask.draw_image(tinyMask, x, y, mode);
```

Will allocate a chunk of memory, render mask into it, then modify mask so that it's a ram image, (not a partial image). 

Then it'll do the draw operation -- this does the same thing to `tinyMask`, so that'll also become a memory image (and get cached by the libvips operation cache).

After you've rendered the blurs, the 1920x1080 memory image gets used as the source for the second pipeline that I guess blurs the edges of the frame.

You can set `VIPS_TRACE` to see what operations are being executed, and which are coming from cache, eg.:

```
$ VIPS_TRACE=1 vips copy k2.jpg x.jpg
vips cache : jpegload filename="k2.jpg" out=((VipsImage*) 0x5584e2ebc010) flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache : copy in=((VipsImage*) 0x5584e2ebc010) out=((VipsImage*) 0x5584e2ebc1a0) -
vips cache+: cast in=((VipsImage*) 0x5584e2ebc1a0) out=((VipsImage*) 0x5584e2ebc330) format=((VipsBandFormat) VIPS_FORMAT_UCHAR) -
vips cache : copy in=((VipsImage*) 0x5584e2ebc330) out=((VipsImage*) 0x5584e2ebc4c0) -
vips cache+: linecache in=((VipsImage*) 0x5584e2ebc7e0) out=((VipsImage*) 0x5584e2ebcb00) tile-height=8 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) -
vips cache+: sequential in=((VipsImage*) 0x5584e2ebc7e0) out=((VipsImage*) 0x5584e2ebc970) tile-height=8 -
vips cache+: extract_area input=((VipsImage*) 0x5584e2ebc970) out=((VipsImage*) 0x5584e2ebcc90) left=0 top=0 width=1450 height=2048 -
vips cache : jpegsave_target in=((VipsImage*) 0x5584e2ebc1a0) target=((VipsTarget*) 0x5584e2ec4bc0) -
```

`vips cache+: ` means that operation has been added to the cache, `vips cache :` means an uncacheable operation, `vips cache*:` means a cache hit.

**Comment by jcupitt** [comment]
I tried to make you a tiny test program:

```C++
/* compile with
 *     g++ -g -Wall blur.cpp `pkg-config vips-cpp --cflags --libs`
 */


#include <vips/vips8>

int
main(int argc, char **argv)
{
        VIPS_INIT(argv[0]);

        const int blurs[][4] = {
                {  10,  20, 100, 100 },
                { 500, 600, 100, 100 },
                {  50, 600,  50,  50 },
        };
        const double blur_radius = 5.0;

        auto image = vips::VImage::new_from_file(argv[1]);

        for (auto &[x, y, w, h] : blurs) {
                auto mask = vips::VImage::black(w, h)
                        .invert()
                        .embed(2 * blur_radius, 
                                2 * blur_radius, 
                                w + 4 * blur_radius, 
                                h + 4 * blur_radius,
                                vips::VImage::option()->set("background", 0)
                         ) 
                        .gaussblur(blur_radius);
                image.draw_image(mask, x, y,    
                        vips::VImage::option()->set("mode", "add"));
        }

        image.write_to_file(argv[2]);

        return 0;
}      
```

If I run it I see:

```
$ VIPS_TRACE=1 ./a.out ~/pics/k2.jpg x.jpg | grep gaussm
vips cache+: gaussmat out=((VipsImage*) 0x55bcb3fae4b0) sigma=5.000000 min-ampl=0.200000 separable=TRUE precision=((VipsPrecision) VIPS_PRECISION_INTEGER) -
vips cache*: gaussmat sigma=5.000000 min-ampl=0.200000 separable=TRUE precision=((VipsPrecision) VIPS_PRECISION_INTEGER) -
vips cache*: gaussmat sigma=5.000000 min-ampl=0.200000 separable=TRUE precision=((VipsPrecision) VIPS_PRECISION_INTEGER) -
```

So the first blur goes into the cache, and it's then reused.  The result looks OK (though not very useful heh).

**Reply by jcupitt** [reply]
Actually, that's a bit confusing. Maybe this is clearer:

```
$ VIPS_TRACE=1 ./a.out ~/pics/k2.jpg x.jpg | grep black
vips cache+: black out=((VipsImage*) 0x55b68c6841a0) width=100 height=100 -
vips cache+: black out=((VipsImage*) 0x55b68c684970) width=1 height=1 bands=1 -
vips cache*: black width=100 height=100 -
vips cache*: black width=1 height=1 bands=1 -
vips cache+: black out=((VipsImage*) 0x55b68c7a0320) width=50 height=50 -
vips cache*: black width=1 height=1 bands=1 -
```

So the first 100x100 is cached, the second is a hit, the 50x50 is cached again.

**Reply by breunigs** [reply]
Aha, `VIPS_TRACE=1` is quite useful to understand what is going on.

In my case I believe the issue is that `auto mask = black(1920, 1080)` is being cached, and the `mask.draw_image(tinyMask, …)` operation writes into that cache. I made libvips save an image of the "freshly" created, empty, black mask for every video frame:
```cpp
// loop over video frames
        auto mask = vips::VImage::black(width, height);
        std::stringstream ss;
        ss << "/tmp/libvips/mask_" << frame << ".jpg";
        mask.jpegsave(ss.str().c_str());
        frame++;

        // blurring happens here
// loop end
```
The image for frame 0 is completely black, as one would expect. And here is the image from frame 10:
![mask_10](https://user-images.githubusercontent.com/307954/166095359-0cd4974d-6ad7-4cab-a013-8ca2ba44966c.jpg)

This is quite surprising to me. It seems that `draw_image` doesn't invalidate the cache of the VImage it's called on. Is this a bug? The "operation cache" explanation seems to indicate that, at least how I understand it: https://www.libvips.org/API/current/How-it-works.md.html I can provide a reproducer for this issue if needed (or the full source to the program if that's helpful, though it's not exactly "hello world").

Regarding my original question/confusion of "why does this trigger when I manually cache the `tinyMask`s" can be answered through chance and cache eviction. Without the manual cache my sample test case runs significantly slower (1m20s instead of 50s), so it stands to reason that it evicts some of the `tinyMask`s and has to re-create them. If this causes the `mask` to be evicted, the issue wouldn't show (or I just didn't notice because it's not that obvious). My follow up question is if I can control that behaviour, since I know things that libvips doesn't – for example, that caching on the `mask` or the actual image is pointless, as they are almost never going to be the same.

**Reply by jcupitt** [reply]
Ah yes, I forgot, you're right. The draw operations don't knock images out of cache because the search for an-image-and-any-downstream-dependencies is potentially quite slow, and it's too much to pay on every operation.

The draw docs say to call copy before you start drawing:

https://www.libvips.org/API/current/libvips-draw.html#libvips-draw.description

You can also call invalidate after drawing, but copy before is faster:

https://www.libvips.org/API/current/VipsImage.html#vips-image-invalidate-all

**Reply by breunigs** [reply]
Ah, I see, makes sense.

As a suggestion for the docs, maybe a reference to that paragraph can be added to every destructive operation. I was looking at the C++ docs (i.e. https://www.libvips.org/API/8.11/cpp/classVImage.html#a598071f631643edbaa2935d09e141204 ) where this isn't mentioned. It didn't occur to me that I would find the general descriptions/trade-offs for draw commands on what looks like the C API reference.

**Reply by jcupitt** [reply]
Yes, the C docs are the main ones. We try to avoid copy-pasting all the docs into every binding because it makes so much extra maintenance work.

The C docs are available as XML, so we could automatically merge them into the doxy output, but no one's found the time yet.
